imbi/endpoints/base.py

```diff
@@ -470,3 +478,133 @@ async def patch(self, *args, **kwargs):
     @require_permission('admin')
     async def post(self, *args, **kwargs):
         await super().post(*args, **kwargs)
+
+
+@dataclasses.dataclass
+class PaginationToken:
+    """Opaque token used in "next" links by the PaginatedRequestMixin"""
+
+    start: datetime.datetime
+    limit: int
+    earliest: datetime.datetime
+
+    def with_start(self, start: datetime.datetime) -> PaginationToken:
+        """Create a new token with a different start value"""
+        return PaginationToken(start=start,
+                               limit=self.limit,
+                               earliest=self.earliest)
+
+    @classmethod
+    def from_header(cls, header_val: str) -> PaginationToken:
+        """Create a token from an opaque header value"""
+        raw_token = common.urlsafe_padded_b64decode(header_val)
+        data = umsgpack.unpackb(raw_token)
+        return PaginationToken(
+            earliest=datetime.datetime.fromisoformat(data['earliest']),
+            limit=data['limit'],
+            start=(datetime.datetime.fromisoformat(data['start']) -
+                   datetime.timedelta.resolution))
+
+    def to_header(self) -> str:
+        """Generate an opaque header value from this token"""
+        return base64.urlsafe_b64encode(
+            umsgpack.packb({
+                'earliest': self.earliest.isoformat(),
+                'limit': self.limit,
+                'start': self.start.isoformat(),
+            })).decode('ascii').rstrip('=')
+
+
+QueryMethod = typing.Callable[[dict[str, typing.Any]],
+                              typing.Coroutine[None, None,
+                                               list[tuple[datetime.datetime,
+                                                          typing.Any]]]]
+"""Type signature of the function called by PaginatedRequestHandler
+
+This is something that looks like:
+
+  async def retrieve_data(
+      params: dict[str, typing.Any]) -> list[tuple[datetime.datetime, dict]:
```

Comment:
Tiny nit: this is missing a `]` at the very end

Reply:
good 👀